builtin/providers/aws/resource_aws_emr_cluster.go

```diff
@@ -0,0 +1,537 @@
+							Optional: true,
+						},
+						"additional_slave_security_groups": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"emr_managed_master_security_group": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"emr_managed_slave_security_group": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"instance_profile": &schema.Schema{
+							Type: schema.TypeString,
+							// TODO: should be ForceNew?
+							Required: true,
+						},
+					},
+				},
+			},
+			"bootstrap_action": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"path": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"args": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Set:      schema.HashString,
+						},
+					},
+				},
+			},
+			"tags": tagsSchema(),
+			"configurations": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"service_role": &schema.Schema{
+				Type: schema.TypeString,
+				// TODO: should be force new?
+				Required: true,
+			},
+			"visible_to_all_users": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+		},
+	}
+}
+
+func resourceAwsEMRClusterCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).emrconn
+
+	log.Printf("[DEBUG] Creating EMR cluster")
+	masterInstanceType := d.Get("master_instance_type").(string)
+	coreInstanceType := masterInstanceType
+	if v, ok := d.GetOk("core_instance_type"); ok {
+		coreInstanceType = v.(string)
+	}
+	coreInstanceCount := d.Get("core_instance_count").(int)
+
+	applications := d.Get("applications").(*schema.Set).List()
+	var userKey, subnet, extraMasterSecGrp, extraSlaveSecGrp, emrMasterSecGrp, emrSlaveSecGrp, instanceProfile string
+
+	if a, ok := d.GetOk("ec2_attributes"); ok {
+		ec2Attributes := a.([]interface{})
+		attributes := ec2Attributes[0].(map[string]interface{})
+		userKey = attributes["key_name"].(string)
+		subnet = attributes["subnet_id"].(string)
+		extraMasterSecGrp = attributes["additional_master_security_groups"].(string)
+		extraSlaveSecGrp = attributes["additional_slave_security_groups"].(string)
+		emrMasterSecGrp = attributes["emr_managed_master_security_group"].(string)
+		emrSlaveSecGrp = attributes["emr_managed_slave_security_group"].(string)
+
+		if len(strings.TrimSpace(attributes["instance_profile"].(string))) != 0 {
+			instanceProfile = strings.TrimSpace(attributes["instance_profile"].(string))
+		}
+	}
+
+	emrApps := expandApplications(applications)
+
+	params := &emr.RunJobFlowInput{
```

Comment:
refactored to omit the empty optionals in 1e8372520ea8c7d17d3fdb830794d27d3893fdc1